### PR TITLE
StringsDicts support

### PR DESCRIPTION
**Pull request by LukasHromadnik**
In combination with SwiftGen 6.3 we are now able to support simple stringsDicts.

Currently we can handle only top level plurals. That means that nested plurals where multiple plural keys are used in one single translation key are not supported.

### CODE REVIEW

File: Sources/ACKLocalizationCore/ACKLocalization.swift [hunk 1]
```diff
@@ -145,13 +145,52 @@ public final class ACKLocalization {
         try mappedValues.forEach { langCode, rows in
             let dirPath = directory + "/" + langCode + ".lproj"
             let filePath = dirPath + "/" + stringsFileName
+            let pluralsPath = dirPath + "/Localizable.stringsDict"
```
Comment (olejnjak) [inline Sources/ACKLocalizationCore/ACKLocalization.swift]: As `Configuration` allows [customization of target strings file](https://github.com/AckeeCZ/ACKLocalization/blob/3b25965e4c6200e3613bda02bcfbfc1171ccdc3c/Sources/ACKLocalizationCore/Model/Configuration.swift#L42), I think that plurals strings dict should respect that as well.

File: ACKLocalization.podspec [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Pod::Spec.new do |s|
   s.name             = 'ACKLocalization'
-  s.version          = '1.0.1'
+  s.version          = '1.1'
```
Comment (olejnjak) [inline ACKLocalization.podspec]: Generally I don't think that you should bump versions in such PRs.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,10 @@
 
 ## master
 
+### 1.1
```
Comment (olejnjak) [inline CHANGELOG.md]: Generally I don't think that you should bump versions in such PRs.

**Review by olejnjak — CHANGES_REQUESTED** [review]
I am also missing some tests regarding this feature 🤔

File: Sources/ACKLocalizationCore/ACKLocalization.swift (hunk 1 above)
Comment (LukasHromadnik) [reply Sources/ACKLocalizationCore/ACKLocalization.swift]: I've added an item to the configuration with name for the stringsDict

File: ACKLocalization.podspec (hunk 2 above)
Comment (LukasHromadnik) [reply ACKLocalization.podspec]: reverted

File: CHANGELOG.md (hunk 3 above)
Comment (LukasHromadnik) [reply CHANGELOG.md]: reverted

**Review by olejnjak — COMMENTED** [review]
Well right now I think we're just missing the tests 🙂

File: Sources/ACKLocalizationCore/ACKLocalization.swift [hunk 4]
```diff
@@ -140,18 +140,66 @@ public final class ACKLocalization {
...
+                // Matches the translation key that contains the predefined plural pattern
+                let patternRange = $0.key.range(of: Constants.pluralPattern, options: .regularExpression),
+                // Loads the base of that translation key
+                let key = $0.key.components(separatedBy: "##").first
+            else { return }
+            
+            let pattern = $0.key[patternRange]
+            
+            // Loads the current plural rule
+            guard let pluralKeyRange = pattern.range(of: Constants.pluralKeyPattern, options: .regularExpression) else { assertionFailure(); return }
```
Comment (fortmarek) [inline Sources/ACKLocalizationCore/ACKLocalization.swift]: I think you can use IMHO nicer syntax `return assertionFailure`
also should not this throw an errror ? 🤔

File: Sources/ACKLocalizationCore/ACKLocalization.swift [hunk 5]
```diff
@@ -140,18 +140,66 @@ public final class ACKLocalization {
...
+            
+            // Using `Range` on the `String` gives us `Substring`
+            // and we need `String`
+            let pluralKey = String(pattern[pluralKeyRange])
+            
+            // Load all translations for the given key
+            var currentTranslations = plurals[key]?.translations ?? []
+            
+            // Check if the plural rule is valid
+            guard let translationKey = PluralRuleKey(rawValue: pluralKey) else { assertionFailure(); return }
```
Comment (fortmarek) [inline Sources/ACKLocalizationCore/ACKLocalization.swift]: same comment as above

**Review by fortmarek — COMMENTED** [review]
two minor comments, otherwise seems good, can't wait for this to land !

File: Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import XCTest
+@testable import ACKLocalizationCore
+
+class AuthAPIServiceMock: AuthAPIServicing {
+    func fetchAccessToken(serviceAccount: ServiceAccount) -> AnyPublisher<AccessToken, RequestError> {
+        Empty().eraseToAnyPublisher()
+    }
+}
+
+class SheetsAPIServiceMock: SheetsAPIServicing {
```
Comment (olejnjak) [inline Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift]: just a nitpick, if `ACKLocalizationPluralsTests` is `final` I would also make this class final

File: Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import Combine
+import XCTest
+@testable import ACKLocalizationCore
+
+class AuthAPIServiceMock: AuthAPIServicing {
```
Comment (olejnjak) [inline Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift]: just a nitpick, if `ACKLocalizationPluralsTests` is `final` I would also make this class final

**Comment by LukasHromadnik** [comment]
I made some changes to the plurals parser. Now we are able to determine if the translation key / plural key is missing, also all those checks throw an `PluralError` so the script won't complete until all plural keys are fine

**Review by olejnjak — APPROVED** [review]

File: Sources/ACKLocalizationCore/Model/PluralError.swift [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import Foundation
+
+public enum PluralError: Error {
+    case missingTranslationKey(String)
+    case missingPluralRule(String)
+    case invalidPluralRule(String)
+}
+
+extension PluralError: Equatable {
```
Comment (fortmarek) [inline Sources/ACKLocalizationCore/Model/PluralError.swift]: `Equatable` should be automatically synthesized, shouldn't it?

File: Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            XCTAssertEqual(plural.translations.count, 1)
+        }
+    }
+    
+    func testMissingTranslationKey() {
+        let rows = [
+            LocRow(key: "##{zero}", value: "zero")
+        ]
+        
+        XCTAssertThrowsError(try ackLocalization.buildPlurals(from: rows)) { error in
```
Comment (fortmarek) [inline Tests/ACKLocalizationCoreTests/ACKLocalization+Plurals.swift:75]: might be nice to add a following extension for `XCTest` (might be handy in the future, too): 
```swift
    func XCTAssertThrowsSpecific<Error: Swift.Error & Equatable, T>(_ closure: @autoclosure () throws -> T, _ error: Error, file: StaticString = #file, line: UInt = #line) {
        do {
            _ = try closure()
        } catch let closureError as Error {
            XCTAssertEqual(error, closureError, file: file, line: line)
        } catch let closureError {
            XCTFail("\(error) is not equal to: \(closureError)", file: file, line: line)
        }
    }
```

Your code can then be changed to something like this:
`XCTAssertThrowsSpecific(try ackLocalization.buildPlurals(from: rows), PluralError.missingTranslationkey(rows[0].key))`

**Review by fortmarek — APPROVED** [review]

**Pull request merged**